ingress controller and dns mgt (#49752)
openshift/openshift-docs: modules/modifying-an-existing-ingress-controller.adoc

// Module included in the following assemblies:
//
// *ingress-controller-dnsmgt.adoc

:_content-type: PROCEDURE
[id="modifying-an-existing-ingress-controller_{context}"]
= Modifying an existing Ingress Controller

As a cluster administrator, you can modify an existing Ingress Controller to manually manage the DNS record lifecycle.

.Prerequisites

* Install the OpenShift CLI (`oc`).
* Log in as a user with `cluster-admin` privileges.

.Procedure

. Modify the chosen `IngressController` to set `dnsManagementPolicy`:

+
[source,terminal]
----

> oc -n openshift-ingress-operator patch ingresscontrollers/<name> --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"type":"LoadBalancerService","loadBalancer":{"dnsManagementPolicy":"Unmanaged", "scope":"<current-scope>"}}}}'

Reviewer: ```suggestion
SCOPE=$(oc -n openshift-ingress-operator get ingresscontroller <name> -o=jsonpath="{.status.endpointPublishingStrategy.loadBalancer.scope}")
oc -n openshift-ingress-operator patch ingresscontrollers/<name> --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"type":"LoadBalancerService","loadBalancer":{"dnsManagementPolicy":"Unmanaged", "scope":"${SCOPE}"}}}}'
```

Author: @candita ptal

Reviewer 2: The suggestion looks fine to me, thanks.

Reviewer 2: @Amrita42 I don't see this in the preview, it is missing the line beginning with SCOPE

Author: Added the line .

// Module included in the following assemblies:
//
// *ingress-controller-dnsmgt.adoc

:_content-type: PROCEDURE
[id="modifying-an-existing-ingress-controller_{context}"]
= Modifying an existing Ingress Controller

> As a cluster administrator, you can modify an existing Ingress Controller to manage the DNS record lifecycle.

Reviewer 2: ```suggestion
As a cluster administrator, you can modify an existing Ingress Controller to manually manage the DNS record lifecycle.
```

Author: ++1 to adding the term "manual" . Thanks .

.Prerequisites

* Install the OpenShift CLI (`oc`).
* Log in as a user with `cluster-admin` privileges.
* A DNS record is created and published.

.Procedure

> . Modify the custom `IngressController` custom resource to set `dnsManagementPolicy`:

Reviewer 2: ```suggestion
. Modify the chosen `IngressController` to set `dnsManagementPolicy`:
```

+
[source,terminal]
----

> oc -n openshift-ingress-operator patch ingresscontrollers/<name> --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"type":"LoadBalancerService","loadBalancer":{"dnsManagementPolicy":"Unmanaged", "scope":"<current-scope>"}}}}'

Reviewer 2: Do they need to change scope for this?
```suggestion
oc -n openshift-ingress-operator patch ingresscontrollers/<name> --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"type":"LoadBalancerService","loadBalancer":{"dnsManagementPolicy":"Unmanaged"}}}}'
```

Author: @thejasn ptal

Reviewer: They don't need to change the scope. Since scope is a required field, we can't really patch `loadBalancer`, we end up with the validation error. 
```
The IngressController "default" is invalid: spec.endpointPublishingStrategy.loadBalancer.scope: Required value
```

Reviewer 2: I guess we have to tell people how to get the `current-scope` then?

Author: @thejasn ptal

Reviewer: I've added a [suggestion](https://github.com/openshift/openshift-docs/pull/49752/files#r974248475), let me know if this is sufficient for the end-user.

----

> . Save the file to apply the changes.

Reviewer 2: There's no file in this section.

+
[source,terminal]
----
oc apply -f <name>.yaml <1>

> ----

Reviewer 2: There's no file in this section.

Author: From my understanding Step 1 , we modify an existing yaml file for the object `IngressController` and then save the file to apply the changes. 
for example ref  https://docs.openshift.com/container-platform/4.11/security/allowing-javascript-access-api-server.html

Reviewer 2: The command on L23 (https://github.com/openshift/openshift-docs/pull/49752/files#diff-575bf2702aeabae30bfbe683d6276cedff8092f80209dcae8d7ed7b42959f83eR23) is run instead of editing a file.

Reviewer 2: @Amrita42 `Step 2` should be deleted.

Author: Thanks @candita implemented.

// Module included in the following assemblies:
//
// *ingress-controller-dnsmgt.adoc

:_content-type: PROCEDURE
[id="modifying-an-existing-ingress-controller_{context}"]
= Modifying an existing Ingress Controller

As a cluster administrator, you can modify an existing Ingress Controller to manage DNS record lifecycle.

.Prerequisites

* Install the OpenShift CLI (`oc`).
* Log in as a user with `cluster-admin` privileges.

> * DNS record created.

Reviewer: ```suggestion
* A DNS record is created and published.
```

Author: I think this pre-req is not required when creating an IG CR

Reviewer: Since we are modifying an existing ing controller to set as `Unmanaged`, we would still need an externally created DNS record.

.Procedure

> . Modify an `IngressController` custom resource to set `dnsManagementPolicy`:

Reviewer: ```suggestion
. Modify the existing custom `IngressController` resource to set `dnsManagementPolicy`:
```

Author: We don't need to mention word, _existing_ as it is already implied. when we tell the user to modify

+
[source,terminal]
----
oc -n openshift-ingress-operator patch ingresscontrollers/<name> --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"loadBalancer":{"dnsManagementPolicy":"Unmanaged"}}}}'
----

+
[NOTE]
====

> Once you update `.spec.endpointPublishingStrategy.loadBalancer.dnsManagementPolicy` field to `Unmanaged`, the existing DNS record must denote the same using the `dnsManagementPolicy` field. You can either retain or delete the DNS record in the cloud provider.

Reviewer: This part I think we don't need to mention, sounds like it confuses the user. I've removed this in the suggestion added below.

Author: done

> If the domain on the Ingress Controller does not match with `dnsConfig.Spec.BaseDomain` , the policy is automatically set to `Unmanaged`.

Author: Have added the sentence thats its AWS specific

====

+

> .Sample Ingress Controller configuration

Reviewer: Why would we need to provide a sample ingress controller yaml under "Modifying an Ingress Controller" section? Seems unnecessary, I think we can only retain it there.

Author: ok done

+
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  namespace: openshift-ingress-operator
  name: <name> <1>
spec:
  domain: <domain> <2>
  endpointPublishingStrategy:
    type: LoadBalancerService
    loadBalancer:
      scope: External <3>
      dnsManagementPolicy: Unmanaged <4>
----
<1> Replace `<name>` with a name for the `IngressController` object.

> <2> Specify the `domain` for the application published by the controller.

Reviewer: ```suggestion
<2> Specify the `domain` based on the DNS record that was created as a prerequisite.
```
